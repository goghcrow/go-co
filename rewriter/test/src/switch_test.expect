// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
	ʂɘʠ "github.com/goghcrow/go-co/seq"
)

func trivalSwitch() (_ ʂɘʠ.Iterator[int]) {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		{
			a := 1
			switch a {
			case 1:
				return ʂɘʠ.Return[int]()

			case 2:
				return ʂɘʠ.Return[int]()

			default:
				return ʂɘʠ.Return[int]()

			}
		}
	}))
}

func TestSwitchInitNameConflict(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			a := 42
			{
				a := a + 1
				switch a {
				case 1:
					assertEqual(t, a, 43)
					return ʂɘʠ.Return[int]()

				case 2:
					assertEqual(t, a, 43)
					return ʂɘʠ.Return[int]()

				default:
					assertEqual(t, a, 43)
					return ʂɘʠ.Return[int]()

				}
			}
		}))
	}
	iter2slice(g())
}

func TestSwitchWithYieldInInit(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
				switch 1 {
				case 1:
					return ʂɘʠ.Return[int]()

				case 2:
					return ʂɘʠ.Return[int]()

				default:
					return ʂɘʠ.Return[int]()

				}
			})
		}))
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1})
}

func TestSwitchWithYieldInInitAndCase(t *testing.T) {
	g := func(a int) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](a, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a {
					case 1:
						return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					case 2:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					default:
						return ʂɘʠ.Bind[int](42, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					}
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Return[int]()
				}))
			})
		}))

	}
	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{1, 1})
	}
	{
		xs := iter2slice(g(2))
		assertEqual(t, xs, []int{2, 2})
	}
	{
		xs := iter2slice(g(3))
		assertEqual(t, xs, []int{3, 42})
	}
}

func TestSwitchWithYieldInInitAndCaseWithoutDefault(t *testing.T) {
	g := func(a int) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](a, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a {
					case 1:
						return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					case 2:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					}
					return ʂɘʠ.Normal[int]()
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Return[int]()
				}))
			})
		}))

	}
	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{1, 1})
	}
	{
		xs := iter2slice(g(2))
		assertEqual(t, xs, []int{2, 2})
	}
	{
		xs := iter2slice(g(3))
		assertEqual(t, xs, []int{3})
	}
}

func TestSwitch(t *testing.T) {
	i := 0
	f := func() { i++ }
	g := func(a, b int) (_ ʂɘʠ.Iterator[string]) {
		return ʂɘʠ.Start[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
			i = 0
			return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
				switch a {
				case 1:
					return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
						switch b {
						case 1:
							return ʂɘʠ.Bind[string]("11", func() ʂɘʠ.Seq[string] {
								return ʂɘʠ.Normal[string]()
							})
						case 2:
							return ʂɘʠ.Bind[string]("12", func() ʂɘʠ.Seq[string] {
								return ʂɘʠ.Normal[string]()
							})
						default:
							return ʂɘʠ.Bind[string]("1?", func() ʂɘʠ.Seq[string] {
								return ʂɘʠ.Normal[string]()
							})
						}
					}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {

						f()
						return ʂɘʠ.Normal[string]()
					}))
				case 2:
					return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
						switch b {
						case 1:
							return ʂɘʠ.Bind[string]("21", func() ʂɘʠ.Seq[string] {
								return ʂɘʠ.Normal[string]()
							})
						case 2:
							return ʂɘʠ.Bind[string]("22", func() ʂɘʠ.Seq[string] {
								return ʂɘʠ.Normal[string]()
							})
						default:
							return ʂɘʠ.Bind[string]("2?", func() ʂɘʠ.Seq[string] {
								return ʂɘʠ.Normal[string]()
							})
						}
					}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {

						f()
						return ʂɘʠ.Normal[string]()
					}))
				}
				return ʂɘʠ.Normal[string]()
			}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {

				f()
				return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
					if i == 1 {
						return ʂɘʠ.Bind[string]("f", func() ʂɘʠ.Seq[string] {
							return ʂɘʠ.Normal[string]()
						})
					} else if i == 2 {
						return ʂɘʠ.Bind[string]("ff", func() ʂɘʠ.Seq[string] {
							return ʂɘʠ.Normal[string]()
						})
					} else {

						panic("unexpected")
					}
				}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
					return ʂɘʠ.Return[string]()
				}))
			}))
		}))

	}

	assertEqual(t, iter2slice(g(1, 1)), []string{"11", "ff"})
	assertEqual(t, iter2slice(g(1, 2)), []string{"12", "ff"})
	assertEqual(t, iter2slice(g(1, 3)), []string{"1?", "ff"})
	assertEqual(t, iter2slice(g(2, 1)), []string{"21", "ff"})
	assertEqual(t, iter2slice(g(2, 2)), []string{"22", "ff"})
	assertEqual(t, iter2slice(g(2, 3)), []string{"2?", "ff"})
	assertEqual(t, iter2slice(g(3, 1)), []string{"f"})
}
