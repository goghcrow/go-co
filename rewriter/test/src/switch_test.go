package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
)

func trivalSwitch() (_ Iter[int]) {
	switch a := 1; a {
	case 1:
		return
	case 2:
		return
	default:
		return
	}
}

func TestSwitchInitNameConflict(t *testing.T) {
	g := func() (_ Iter[int]) {
		a := 42
		switch a := a + 1; a {
		case 1:
			assertEqual(t, a, 43)
			return
		case 2:
			assertEqual(t, a, 43)
			return
		default:
			assertEqual(t, a, 43)
			return
		}
	}
	iter2slice(g())
}

func TestSwitchWithYieldInInit(t *testing.T) {
	g := func() (_ Iter[int]) {
		switch Yield(1); 1 {
		case 1:
			return
		case 2:
			return
		default:
			return
		}
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1})
}

func TestSwitchWithYieldInInitAndCase(t *testing.T) {
	g := func(a int) (_ Iter[int]) {
		switch Yield(a); a {
		case 1:
			Yield(1)
		case 2:
			Yield(2)
		default:
			Yield(42)
		}
		return
	}
	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{1, 1})
	}
	{
		xs := iter2slice(g(2))
		assertEqual(t, xs, []int{2, 2})
	}
	{
		xs := iter2slice(g(3))
		assertEqual(t, xs, []int{3, 42})
	}
}

func TestSwitchWithYieldInInitAndCaseWithoutDefault(t *testing.T) {
	g := func(a int) (_ Iter[int]) {
		switch Yield(a); a {
		case 1:
			Yield(1)
		case 2:
			Yield(2)
		}
		return
	}
	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{1, 1})
	}
	{
		xs := iter2slice(g(2))
		assertEqual(t, xs, []int{2, 2})
	}
	{
		xs := iter2slice(g(3))
		assertEqual(t, xs, []int{3})
	}
}

func TestSwitch(t *testing.T) {
	i := 0
	f := func() { i++ }
	g := func(a, b int) (_ Iter[string]) {
		switch i = 0; a {
		case 1:
			switch b {
			case 1:
				Yield("11")
			case 2:
				Yield("12")
			default:
				Yield("1?")
			}
			f()
		case 2:
			switch b {
			case 1:
				Yield("21")
			case 2:
				Yield("22")
			default:
				Yield("2?")
			}
			f()
		}
		f()
		if i == 1 {
			Yield("f")
		} else if i == 2 {
			Yield("ff")
		} else {
			panic("unexpected")
		}
		return
	}

	assertEqual(t, iter2slice(g(1, 1)), []string{"11", "ff"})
	assertEqual(t, iter2slice(g(1, 2)), []string{"12", "ff"})
	assertEqual(t, iter2slice(g(1, 3)), []string{"1?", "ff"})
	assertEqual(t, iter2slice(g(2, 1)), []string{"21", "ff"})
	assertEqual(t, iter2slice(g(2, 2)), []string{"22", "ff"})
	assertEqual(t, iter2slice(g(2, 3)), []string{"2?", "ff"})
	assertEqual(t, iter2slice(g(3, 1)), []string{"f"})
}
