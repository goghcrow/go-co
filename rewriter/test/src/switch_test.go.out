//go:build !co

// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	ʂɘʠ "github.com/goghcrow/go-co/seq"
	"testing"
)

func trivalSwitch() (_ ʂɘʠ.Iterator[int]) {
	return ʂɘʠ.Start[int](
		ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
			{

				a := 1
				switch a {
				case 1:
					return ʂɘʠ.Return[int]()

				case 2:
					return ʂɘʠ.Return[int]()

				default:
					return ʂɘʠ.Return[int]()

				}
			}
		}),
	)
}

func TestSwitchInitNameConflict(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {

				a := 42
				{
					a := a + 1
					switch a {
					case 1:
						assertEqual(t, a, 43)
						return ʂɘʠ.Return[int]()

					case 2:
						assertEqual(t, a, 43)
						return ʂɘʠ.Return[int]()

					default:
						assertEqual(t, a, 43)
						return ʂɘʠ.Return[int]()

					}
				}
			}),
		)
	}
	iter2slice(g())
}

func TestSwitchWithYieldInInit(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
				switch 1 {
				case 1:
					return ʂɘʠ.Return[int]()

				case 2:
					return ʂɘʠ.Return[int]()

				default:
					return ʂɘʠ.Return[int]()

				}
			}),
		)
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1})
}

func TestSwitchWithYieldInInitAndCase(t *testing.T) {
	g := func(a int) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](a, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a {
					case 1:
						return ʂɘʠ.Bind[int](1,
							ʂɘʠ.Normal[int],
						)
					case 2:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Normal[int],
						)
					default:
						return ʂɘʠ.Bind[int](42,
							ʂɘʠ.Normal[int],
						)
					}
				}),
					ʂɘʠ.Return[int](),
				)
			})
		}))

	}
	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{1, 1})
	}
	{
		xs := iter2slice(g(2))
		assertEqual(t, xs, []int{2, 2})
	}
	{
		xs := iter2slice(g(3))
		assertEqual(t, xs, []int{3, 42})
	}
}

func TestSwitchWithYieldInInitAndCaseWithoutDefault(t *testing.T) {
	g := func(a int) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](a, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a {
					case 1:
						return ʂɘʠ.Bind[int](1,
							ʂɘʠ.Normal[int],
						)
					case 2:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Normal[int],
						)
					}
					return ʂɘʠ.Normal[int]()
				}),
					ʂɘʠ.Return[int](),
				)
			})
		}))

	}
	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{1, 1})
	}
	{
		xs := iter2slice(g(2))
		assertEqual(t, xs, []int{2, 2})
	}
	{
		xs := iter2slice(g(3))
		assertEqual(t, xs, []int{3})
	}
}

func TestSwitch(t *testing.T) {
	i := 0
	f := func() { i++ }
	g := func(a, b int) (_ ʂɘʠ.Iterator[string]) {
		return ʂɘʠ.Start[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
			i = 0
			return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
				switch a {
				case 1:
					return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
						switch b {
						case 1:
							return ʂɘʠ.Bind[string]("11",
								ʂɘʠ.Normal[string],
							)
						case 2:
							return ʂɘʠ.Bind[string]("12",
								ʂɘʠ.Normal[string],
							)
						default:
							return ʂɘʠ.Bind[string]("1?",
								ʂɘʠ.Normal[string],
							)
						}
					}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {

						f()
						return ʂɘʠ.Normal[string]()
					}))
				case 2:
					return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
						switch b {
						case 1:
							return ʂɘʠ.Bind[string]("21",
								ʂɘʠ.Normal[string],
							)
						case 2:
							return ʂɘʠ.Bind[string]("22",
								ʂɘʠ.Normal[string],
							)
						default:
							return ʂɘʠ.Bind[string]("2?",
								ʂɘʠ.Normal[string],
							)
						}
					}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {

						f()
						return ʂɘʠ.Normal[string]()
					}))
				}
				return ʂɘʠ.Normal[string]()
			}), ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {

				f()
				return ʂɘʠ.Combine[string](ʂɘʠ.Delay[string](func() ʂɘʠ.Seq[string] {
					if i == 1 {
						return ʂɘʠ.Bind[string]("f",
							ʂɘʠ.Normal[string],
						)
					} else if i == 2 {
						return ʂɘʠ.Bind[string]("ff",
							ʂɘʠ.Normal[string],
						)
					} else {

						panic("unexpected")
					}
				}),
					ʂɘʠ.Return[string](),
				)
			}))
		}))

	}

	assertEqual(t, iter2slice(g(1, 1)), []string{"11", "ff"})
	assertEqual(t, iter2slice(g(1, 2)), []string{"12", "ff"})
	assertEqual(t, iter2slice(g(1, 3)), []string{"1?", "ff"})
	assertEqual(t, iter2slice(g(2, 1)), []string{"21", "ff"})
	assertEqual(t, iter2slice(g(2, 2)), []string{"22", "ff"})
	assertEqual(t, iter2slice(g(2, 3)), []string{"2?", "ff"})
	assertEqual(t, iter2slice(g(3, 1)), []string{"f"})
}
